File: scripts/system-startup.py
import subprocess
import time
import os
import sys
import signal
import urllib.request
import socket

# Configuration
NODE_ENV = "production"
ROOT = os.getcwd()
SERVICES = [
    {
        "name": "Socket Server",
        "cwd": "apps/api-gateway",
        "cmd": ["npx", "tsx", "-r", "tsconfig-paths/register", "services/socket.ts"],
        "health": "http://127.0.0.1:3011/health"
    },
    {
        "name": "Frontend",
        "cwd": "apps/frontend",
        "cmd": ["npx", "tsx", "-r", "tsconfig-paths/register", "src/index.ts"],
        "health": "http://localhost:3004"
    }
]

processes = []

def cleanup(sig=None, frame=None):
    print("\n🛑 Shutting down MultiAgent...")
    for proc in processes:
        try:
            # On Windows, we need to be aggressive with taskkill /T /F
            subprocess.run(["taskkill", "/F", "/T", "/PID", str(proc.pid)], 
                          stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except:
            proc.kill()
    sys.exit(0)

signal.signal(signal.SIGINT, cleanup)

def is_port_open(port):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        return s.connect_ex(('127.0.0.1', port)) == 0

def kill_port(port):
    if os.name == 'nt':
        try:
            # Surgical port killing via PowerShell
            cmd = f"Get-NetTCPConnection -LocalPort {port} -ErrorAction SilentlyContinue | Select-Object -ExpandProperty OwningProcess | ForEach-Object {{ Stop-Process -Id $_ -Force -ErrorAction SilentlyContinue }}"
            subprocess.run(["powershell", "-Command", cmd], capture_output=True)
        except:
            pass

def wait_for_health(name, url, timeout=60):
    print(f"⏳ Waiting for {name} health: {url}")
    start = time.time()
    while time.time() - start < timeout:
        try:
            with urllib.request.urlopen(url, timeout=2) as response:
                if response.status == 200:
                    print(f"\n✅ {name} is healthy!")
                    return True
        except:
            pass
        sys.stdout.write('.')
        sys.stdout.flush()
        time.sleep(2)
    print(f"\n❌ {name} health check timed out")
    return False

def boot():
    print("\n=== 🔮 MultiAgent Python Bootstrapper ===\n")
    
    # 1. Clean ports
    kill_port(3011)
    kill_port(3004)
    
    # 2. Start Socket Server
    s = SERVICES[0]
    print(f"🚀 Starting {s['name']}...")
    env = os.environ.copy()
    env["NODE_ENV"] = NODE_ENV
    env["SERVICE_NAME"] = s["name"]
    
    # Force absolute npx for safety
    npx_bin = "npx.cmd" if os.name == 'nt' else 'npx'
    
    proc = subprocess.Popen(
        [npx_bin, "tsx", "-r", "tsconfig-paths/register", "services/socket.ts"], 
        cwd=os.path.join(ROOT, s["cwd"]),
        env=env,
        shell=True # Required for .cmd binaries and space handling on Windows
    )
    processes.append(proc)
    
    if not wait_for_health(s["name"], s["health"]):
        cleanup()
        
    # 3. Start Frontend
    f = SERVICES[1]
    print(f"🚀 Starting {f['name']}...")
    proc_f = subprocess.Popen(
        [npx_bin, "tsx", "-r", "tsconfig-paths/register", "src/index.ts"],
        cwd=os.path.join(ROOT, f["cwd"]),
        env=env,
        shell=True
    )
    processes.append(proc_f)
    
    print("\n🎉 SYSTEM FULLY OPERATIONAL\n")
    
    # Keep main alive
    try:
        while True:
            if proc.poll() is not None:
                print(f"❌ {s['name']} exited unexpectedly")
                cleanup()
            if proc_f.poll() is not None:
                print(f"❌ {f['name']} exited unexpectedly")
                cleanup()
            time.sleep(5)
    except KeyboardInterrupt:
        cleanup()

if __name__ == "__main__":
    boot()

File: scripts/system-startup.mjs
import { spawn, execSync, spawnSync } from 'node:child_process';
import http from 'node:http';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

process.env.NODE_ENV = 'production';

// 🔥 HARD GUARD: Prevent multiple instances of the orchestrator itself
if (process.env.MULTIAGENT_ORCHESTRATOR_ACTIVE === 'true') {
  console.log('⚠️ Orchestrator already active. Skipping duplicate start.');
  process.exit(0);
}
process.env.MULTIAGENT_ORCHESTRATOR_ACTIVE = 'true';

// Using relative path for binary to bypass Windows path length/mangling issues in shell
const TSX = 'node_modules\\.bin\\tsx.cmd'; 

console.log('\n=== 🚀 MultiAgent Hardened Boot (ESM v3) ===\n');

const processes = [];

function startProcess(name, appPath, args, customBin, retryCount = 0) {
  console.log(`🚀 Starting ${name} ${retryCount > 0 ? `(Retry ${retryCount})...` : '...'}`);
  
  let bin;
  if (customBin) {
    bin = customBin;
  } else {
    const relRoot = appPath.split('/').map(() => '..').join('/');
    bin = path.join(relRoot, TSX);
  }
  
  const proc = spawn(bin, args, { 
    cwd: path.join(process.cwd(), appPath),
    stdio: 'inherit',
    shell: true,
    env: { ...process.env, SERVICE_NAME: name, NODE_ENV: 'development' }
  });

  proc.on('exit', (code) => { 
    if (code !== 0 && code !== null) { 
        console.error(`\n❌ ${name} exited with code ${code}`); 
        shutdown(); 
    } 
  });
  processes.push(proc);
}

async function waitForHealth(name, url, timeout) {
  console.log(`⏳ Waiting for ${name} health: ${url}`);
  const start = Date.now();
  while (Date.now() - start < timeout) {
    try {
        const ok = await new Promise((resolve) => {
            const req = http.get(url, (res) => { 
                if (res.statusCode < 200 || res.statusCode >= 400) {
                    process.stdout.write(`(${res.statusCode})`);
                }
                resolve(res.statusCode >= 200 && res.statusCode < 400); 
            });
            req.on('error', () => resolve(false));
            req.setTimeout(10000, () => { req.destroy(); resolve(false); });
            req.end();
        });
        if (ok) { console.log(`\n✅ ${name} is healthy!`); return; }
    } catch (_) {}
    process.stdout.write('.');
    await new Promise((r) => setTimeout(r, 2000));
  }
  throw new Error(`\n❌ ${name} failed health check`);
}

function killPort(port) {
  console.log(`Sweep cleaning port ${port}...`);
  try {
    if (os.platform() === 'win32') {
        const pwsh = 'C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe';
        const killCmd = `"${pwsh}" -Command "Get-NetTCPConnection -LocalPort ${port} -ErrorAction SilentlyContinue | ForEach-Object { Stop-Process -Id $_.OwningProcess -Force -ErrorAction SilentlyContinue }"`;
        spawnSync(killCmd, { stdio: 'ignore', shell: true });
    }
  } catch (_) {}
}

function shutdown() {
  console.log('\n🛑 Shutting down MultiAgent...\n');
  const pwsh = 'C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe';
  for (const proc of processes) { 
    try { 
        if (os.platform() === 'win32') {
            spawnSync(`"${pwsh}" -Command "Stop-Process -Id ${proc.pid} -Force -ErrorAction SilentlyContinue"`, { stdio: 'ignore', shell: true }); 
        } else {
            proc.kill('SIGTERM');
        }
    } catch (_) {} 
  }
  process.exit(1);
}

process.on('SIGINT', shutdown);

(async () => {
  try {
    killPort(3004); killPort(3011); killPort(9091); killPort(3006); killPort(4000);

    console.log('🔍 Health Check: Core Services...');
    
    // Check if TSX is available
    const rootBin = path.join(process.cwd(), TSX);
    
    // 🚀 1. Socket Server
    startProcess('Socket Server', 'apps/api-gateway', ['--tsconfig', 'tsconfig.json', 'services/socket.ts']);
    await waitForHealth('Socket Server', 'http://127.0.0.1:3011/health', 120000);

    // 🤖 2. Build Worker
    startProcess('Build Worker', 'apps/worker', ['--tsconfig', 'tsconfig.json', 'src/build-worker.ts']);
    
    // 🌐 3. Next.js Frontend
    startProcess('Frontend', 'apps/frontend', ['run', 'dev'], 'pnpm');
    await waitForHealth('Frontend', 'http://127.0.0.1:3004', 180000);

    console.log('\n🎉 SYSTEM FULLY OPERATIONAL\n');
    process.stdin.resume();
  } catch (err) { 
    console.error(`\n💥 BOOT FAILURE: ${err.message}`); 
    shutdown(); 
  }
})();
